crates/net/network/src/discovery.rs

```diff
@@ -57,18 +70,21 @@ impl Discovery {
     /// This will spawn the [`reth_discv4::Discv4Service`] onto a new task and establish a listener
     /// channel to receive all discovered nodes.
     pub async fn new(
-        discovery_addr: SocketAddr,
+        discovery_v4_addr: SocketAddr,
         sk: SecretKey,
         discv4_config: Option<Discv4Config>,
+        discv5_config: Option<reth_discv5::Config>, // contains discv5 listen address
         dns_discovery_config: Option<DnsDiscoveryConfig>,
     ) -> Result<Self, NetworkError> {
         // setup discv4
-        let local_enr = NodeRecord::from_secret_key(discovery_addr, &sk);
+        let local_enr = NodeRecord::from_secret_key(discovery_v4_addr, &sk);
         let (discv4, discv4_updates, _discv4_service) = if let Some(disc_config) = discv4_config {
             let (discv4, mut discv4_service) =
-                Discv4::bind(discovery_addr, local_enr, sk, disc_config).await.map_err(|err| {
-                    NetworkError::from_io_error(err, ServiceKind::Discovery(discovery_addr))
-                })?;
+                Discv4::bind(discovery_v4_addr, local_enr, sk, disc_config).await.map_err(
```

Comment:
I get your point that this setup is now even more convoluted.

this is okay for this PR.
but we could make this a bit nicer separately and run the two setups concurrently.
perhaps a match over both configs would be "more" ergonomic here

Reply:
on the down side, when running both versions, this means a significant overhead for the reth node. it will be having double up discovery connections, once on each version, for one unique mempool peer. on the bright side, from an ethereum wide perspective, it will significantly help peer discovery, since finding specific peers becomes easier the more well connected nodes exist in the network. for example DAS will prosper form this.

not sure if it improves readability, since the two versions now have no dependency on each other, i.e. discv4 is run the same with discv5 as without discv5 and vv.